docs/proposals/20190708-MachineConfig.md

```diff
@@ -0,0 +1,293 @@
+        dhcp: true
+install:
+  wipe: false
+  force: true
+  boot:
+    device: /dev/sda
+    size: 1024000000
+    kernel: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/vmlinuz
+    initramfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/initramfs.xz
+  root:
+    device: /dev/sda
+    size: 2048000000
+    rootfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/rootfs.tar.gz
+  data:
+    device: /dev/sda
+    size: 4096000000
+```
+
+#### Worker:
+```
+#!talos
+version: ""
+security: null
+services:
+  init:
+    cni: flannel
+  kubeadm:
+    configuration: |
+      apiVersion: kubeadm.k8s.io/v1beta1
+      kind: JoinConfiguration
+      discovery:
+        bootstrapToken:
+          token: '{{ .KubeadmTokens.BootstrapToken }}'
+          unsafeSkipCAVerification: true
+          apiServerEndpoint: {{ index .MasterIPs 0 }}:443
+      nodeRegistration:
+        taints: []
+        kubeletExtraArgs:
+          node-labels: ""
+          feature-gates: ExperimentalCriticalPodAnnotation=true
+  trustd:
+    token: '{{ .TrustdInfo.Token }}'
+    endpoints: [ {{ .Endpoints }} ]
+networking:
+  os:
+    devices:
+      - interface: lo
+        cidr: x.x.x.x/32
+      - interface: eth0
+        dhcp: true
+install:
+  wipe: false
+  force: true
+  boot:
+    device: /dev/sda
+    size: 1024000000
+    kernel: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/vmlinuz
+    initramfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/initramfs.xz
+  root:
+    device: /dev/sda
+    size: 2048000000
+    rootfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/rootfs.tar.gz
+  data:
+    device: /dev/sda
+    size: 4096000000
+```
+
+
+## Proposal
+
+As you can see, there are quite a few fields that are similar among all node types, while the main difference between them is the kubeadm templates. Given all the proper info, we should be able to generate the kubeadm configuration on the fly depending on node type and abstract the need to know all the different options away. I'm proposing a machine configuration that looks like this:
+
+
+```
+#!talos
+version: "v0.1.0"
+security:
+  os:
+    ca:
+      crt: "{{ .Certs.OsCert }}"
+      key: "{{ .Certs.OsKey }}"
+    token: abc.1234
+  kubernetes:
+    ca:
+      crt: "{{ .Certs.K8sCert }}"
+      key: "{{ .Certs.K8sKey }}"
+    token: abc.1234
+machine:
+  nodeType: init, controlplane, worker
+  masters: [x.x.x.x, y.y.y.y, z.z.z.z]
+  kubeletExtraArgs: {}
+cluster:
+  clusterName: {{ .ClusterName }}
+  controlPlane:
+    apiServerExtraArgs: {}
+    controllerManagerExtraArgs: {}
+    schedulerExtraArgs: {}
+    networking:
+      dnsDomain: {{ .ServiceDomain }}
+      podSubnet: {{ index .PodNet 0 }}
+      serviceSubnet: {{ index .ServiceNet 0 }}
+networking:
+  os:
+    devices:
+      - interface: lo
+        cidr: x.x.x.x/32
+      - interface: eth0
+        dhcp: true
+install:
+  wipe: false
+  force: true
```

Comment:
What does force mean?

Reply:
This could be named better, but it means that the filesystem will be forcefully formatted, vs a wipe where the block device is zeroed out.

Reply:
This essentially translates into the `-f` flag to `mkfs.xfs`.

Reply:
In that case, "force" should probably be moved to be a volume-level option, as "wipe" there.

"wipe" here, then becomes ambiguous.  Does that mean to wipe the partition table?  If so, does that mean for each device listed amongst the volume "device"s?

We had discussed the precept that if Talos is on a disk, it "owns" that disk, meaning the disk and all its partitions.  Leaving "wipe" (partitions) (and to a lesser extent the "wipe" (filesystems)) seems to suggest that to not be true.

I'm not saying it's necessarily a problem to include the option, but it makes the logic more complicated.

My default assumption would be to sum up the list of "devices" for whom "wipe" is enabled, determine which are non-partition specifications (/dev/xvda vs /dev/xvda3 or /dev/nvme0n1 vs /dev/nvme0n1p3), and repartition those disks accordingly.

However, even then, it's a bit hairy.  Just the name 'device' leaves open to discussion whether that _should_ be a disk or a partition.  I think I would just plan for that to be "disk" rather than "device," and always expect to "own" the disk by repartitioning it as needed.

Leaving partitions open to the user is just problematic.  What if there is a partition clash?  What if the existing partition table is dos and there are already 4 primary partitions?

CoreOS did this well:  it expected a single disk which would be unilaterally repartitioned into its own schema.  Then, if you wanted to add additional storage volumes, you did so separately, specifying the parameters to do so.

Reply:
Makes sense. These options might be better left out for the "owned" disk, but available for "extra" disks.